File: app/src/main/java/dev/ragnarok/fenrir/fragment/VKPhotosFragment.java
package dev.ragnarok.fenrir.fragment;

import static android.app.Activity.RESULT_OK;
import static dev.ragnarok.fenrir.util.Objects.nonNull;
import static dev.ragnarok.fenrir.util.Utils.nonEmpty;

import android.Manifest;
import android.content.Intent;
import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.Menu;
import android.view.MenuInflater;
import android.view.MenuItem;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import android.widget.Toast;

import androidx.activity.result.ActivityResultLauncher;
import androidx.activity.result.contract.ActivityResultContracts;
import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.recyclerview.widget.GridLayoutManager;
import androidx.recyclerview.widget.RecyclerView;
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout;

import com.google.android.material.floatingactionbutton.FloatingActionButton;

import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

import dev.ragnarok.fenrir.Extra;
import dev.ragnarok.fenrir.R;
import dev.ragnarok.fenrir.activity.ActivityFeatures;
import dev.ragnarok.fenrir.activity.PhotosActivity;
import dev.ragnarok.fenrir.adapter.BigVkPhotosAdapter;
import dev.ragnarok.fenrir.dialog.ImageSizeAlertDialog;
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment;
import dev.ragnarok.fenrir.listener.EndlessRecyclerOnScrollListener;
import dev.ragnarok.fenrir.listener.OnSectionResumeCallback;
import dev.ragnarok.fenrir.listener.PicassoPauseOnScrollListener;
import dev.ragnarok.fenrir.model.LocalPhoto;
import dev.ragnarok.fenrir.model.Owner;
import dev.ragnarok.fenrir.model.ParcelableOwnerWrapper;
import dev.ragnarok.fenrir.model.Photo;
import dev.ragnarok.fenrir.model.PhotoAlbum;
import dev.ragnarok.fenrir.model.TmpSource;
import dev.ragnarok.fenrir.model.wrappers.SelectablePhotoWrapper;
import dev.ragnarok.fenrir.mvp.core.IPresenterFactory;
import dev.ragnarok.fenrir.mvp.presenter.VkPhotosPresenter;
import dev.ragnarok.fenrir.mvp.view.IVkPhotosView;
import dev.ragnarok.fenrir.place.PlaceFactory;
import dev.ragnarok.fenrir.settings.Settings;
import dev.ragnarok.fenrir.upload.Upload;
import dev.ragnarok.fenrir.util.AppPerms;
import dev.ragnarok.fenrir.util.ViewUtils;

public class VKPhotosFragment extends BaseMvpFragment<VkPhotosPresenter, IVkPhotosView>
        implements BigVkPhotosAdapter.PhotosActionListener, BigVkPhotosAdapter.UploadActionListener, IVkPhotosView {

    private static final String TAG = VKPhotosFragment.class.getSimpleName();
    private final ActivityResultLauncher<Intent> requestUploadPhoto = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == RESULT_OK) {
                    ArrayList<LocalPhoto> photos = result.getData().getParcelableArrayListExtra(Extra.PHOTOS);
                    if (nonEmpty(photos)) {
                        onPhotosForUploadSelected(photos);
                    }
                }
            });
    private final AppPerms.doRequestPermissions requestReadPermission = AppPerms.requestPermissions(this,
            new String[]{Manifest.permission.READ_EXTERNAL_STORAGE},
            () -> callPresenter(VkPhotosPresenter::fireReadStoragePermissionChanged));
    private final AppPerms.doRequestPermissions requestReadPermissionForLoadDownload = AppPerms.requestPermissions(this,
            new String[]{Manifest.permission.READ_EXTERNAL_STORAGE},
            () -> callPresenter(VkPhotosPresenter::loadDownload));
    private final ActivityResultLauncher<Intent> requestPhotoUpdate = registerForActivityResult(new ActivityResultContracts.StartActivityForResult(),
            result -> {
                if (result.getResultCode() == RESULT_OK && result.getData() != null && result.getData().getExtras() != null) {
                    postPresenterReceive(p -> p.updateInfo(result.getData().getExtras().getInt(Extra.POSITION), result.getData().getExtras().getLong(Extra.PTR)));
                }
            });
    private SwipeRefreshLayout mSwipeRefreshLayout;
    private BigVkPhotosAdapter mAdapter;
    private TextView mEmptyText;
    private FloatingActionButton mFab;
    private String mAction;
    private RecyclerView mRecyclerView;

    public static Bundle buildArgs(int accountId, int ownerId, int albumId, String action) {
        Bundle args = new Bundle();
        args.putInt(Extra.ACCOUNT_ID, accountId);
        args.putInt(Extra.OWNER_ID, ownerId);
        args.putInt(Extra.ALBUM_ID, albumId);
        args.putString(Extra.ACTION, action);
        return args;
    }

    public static VKPhotosFragment newInstance(Bundle args) {
        VKPhotosFragment fragment = new VKPhotosFragment();
        fragment.setArguments(args);
        return fragment;
    }

    public static VKPhotosFragment newInstance(int accountId, int ownerId, int albumId, String action) {
        return newInstance(buildArgs(accountId, ownerId, albumId, action));
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        mAction = requireArguments().getString(Extra.ACTION, ACTION_SHOW_PHOTOS);
        setHasOptionsMenu(true);
    }

    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
        View root = inflater.inflate(R.layout.fragment_photo_gallery, container, false);
        ((AppCompatActivity) requireActivity()).setSupportActionBar(root.findViewById(R.id.toolbar));

        int columnCount = getResources().getInteger(R.integer.local_gallery_column_count);
        RecyclerView.LayoutManager manager = new GridLayoutManager(requireActivity(), columnCount);

        mSwipeRefreshLayout = root.findViewById(R.id.refresh);
        mSwipeRefreshLayout.setOnRefreshListener(() -> callPresenter(VkPhotosPresenter::fireRefresh));

        ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout);

        mRecyclerView = root.findViewById(R.id.list);
        mRecyclerView.setLayoutManager(manager);
        mRecyclerView.addOnScrollListener(new PicassoPauseOnScrollListener(TAG));
        mRecyclerView.addOnScrollListener(new EndlessRecyclerOnScrollListener() {
            @Override
            public void onScrollToLastElement() {
                callPresenter(VkPhotosPresenter::fireScrollToEnd);
            }
        });

        mEmptyText = root.findViewById(R.id.empty);

        mFab = root.findViewById(R.id.fr_photo_gallery_attach);
        mFab.setOnClickListener(v -> onFabClicked());

        mAdapter = new BigVkPhotosAdapter(requireActivity(), Collections.emptyList(), Collections.emptyList(), TAG);
        mAdapter.setPhotosActionListener(this);
        mAdapter.setUploadActionListener(this);
        mRecyclerView.setAdapter(mAdapter);
        return root;
    }

    private void resolveEmptyTextVisibility() {
        if (nonNull(mEmptyText) && nonNull(mAdapter)) {
            mEmptyText.setVisibility(mAdapter.getItemCount() == 0 ? View.VISIBLE : View.GONE);
        }
    }

    private void resolveFabVisibility(boolean anim, boolean show) {
        if (!isAdded() || mFab == null) return;

        if (mFab.isShown() && !show) {
            mFab.hide();
        }

        if (!mFab.isShown() && show) {
            mFab.show();
        }
    }

    private void onFabClicked() {
        if (isSelectionMode()) {
            callPresenter(VkPhotosPresenter::fireSelectionCommitClick);
        } else {
            callPresenter(VkPhotosPresenter::fireAddPhotosClick);
        }
    }

    private boolean isSelectionMode() {
        return ACTION_SELECT_PHOTOS.equals(mAction);
    }

    private void onPhotosForUploadSelected(@NonNull List<LocalPhoto> photos) {
        ImageSizeAlertDialog.showUploadPhotoSizeIfNeed(requireActivity(), size -> doUploadPhotosToAlbum(photos, size));
    }

    private void doUploadPhotosToAlbum(@NonNull List<LocalPhoto> photos, int size) {
        callPresenter(p -> p.firePhotosForUploadSelected(photos, size));
    }

    @Override
    public void onResume() {
        super.onResume();
        new ActivityFeatures.Builder()
                .begin()
                .setHideNavigationMenu(false)
                .setBarsColored(requireActivity(), true)
                .build()
                .apply(requireActivity());
    }

    @Override
    public void onDestroyView() {
        mAdapter.cleanup();
        super.onDestroyView();
    }

    @Override
    public void onPhotoClick(BigVkPhotosAdapter.PhotoViewHolder holder, SelectablePhotoWrapper wrapper) {
        if (isSelectionMode()) {
            callPresenter(p -> p.firePhotoSelectionChanged(wrapper));
            mAdapter.updatePhotoHoldersSelectionAndIndexes();
        } else {
            callPresenter(p -> p.firePhotoClick(wrapper));
        }
    }

    @Override
    public void onUploadRemoveClicked(Upload upload) {
        callPresenter(p -> p.fireUploadRemoveClick(upload));
    }

    @Override
    public void displayData(List<SelectablePhotoWrapper> photos, List<Upload> uploads) {
        if (nonNull(mAdapter)) {
            mAdapter.setData(BigVkPhotosAdapter.DATA_TYPE_UPLOAD, uploads);
            mAdapter.setData(BigVkPhotosAdapter.DATA_TYPE_PHOTO, photos);
            mAdapter.notifyDataSetChanged();
            resolveEmptyTextVisibility();
        }
    }

    @Override
    public void notifyDataSetChanged() {
        if (nonNull(mAdapter)) {
            mAdapter.notifyDataSetChanged();
            resolveEmptyTextVisibility();
        }
    }

    @Override
    public void notifyPhotosAdded(int position, int count) {
        if (nonNull(mAdapter)) {
            mAdapter.notifyItemRangeInserted(position, count, BigVkPhotosAdapter.DATA_TYPE_PHOTO);
            resolveEmptyTextVisibility();
        }
    }

    @Override
    public void displayRefreshing(boolean refreshing) {
        if (nonNull(mSwipeRefreshLayout)) {
            mSwipeRefreshLayout.setRefreshing(refreshing);
        }
    }

    @Override
    public void notifyUploadAdded(int position, int count) {
        if (nonNull(mAdapter)) {
            mAdapter.notifyItemRangeInserted(position, count, BigVkPhotosAdapter.DATA_TYPE_UPLOAD);
            resolveEmptyTextVisibility();
        }
    }

    @Override
    public void notifyUploadRemoved(int index) {
        if (nonNull(mAdapter)) {
            mAdapter.notifyItemRemoved(index, BigVkPhotosAdapter.DATA_TYPE_UPLOAD);
            resolveEmptyTextVisibility();
        }
    }

    @Override
    public void setButtonAddVisible(boolean visible, boolean anim) {
        if (nonNull(mFab)) {
            resolveFabVisibility(anim, visible);
        }
    }

    @Override
    public void notifyUploadItemChanged(int index) {
        if (nonNull(mAdapter)) {
            mAdapter.notifyItemChanged(index, BigVkPhotosAdapter.DATA_TYPE_UPLOAD);
        }
    }

    @Override
    public void notifyUploadProgressChanged(int id, int progress) {
        if (nonNull(mAdapter)) {
            mAdapter.updateUploadHoldersProgress(id, true, progress);
        }
    }

    @Override
    public void displayGallery(int accountId, int albumId, int ownerId, @NonNull TmpSource source, int position) {
        PlaceFactory.getPhotoAlbumGalleryPlace(accountId, albumId, ownerId, source, position, false, Settings.get().other().isInvertPhotoRev()).tryOpenWith(requireActivity());
    }

    @Override
    public void displayGalleryUnSafe(int accountId, int albumId, int ownerId, long parcelNativePointer, int position) {
        PlaceFactory.getPhotoAlbumGalleryPlace(accountId, albumId, ownerId, parcelNativePointer, position, false, Settings.get().other().isInvertPhotoRev()).setActivityResultLauncher(requestPhotoUpdate).tryOpenWith(requireActivity());
    }

    @Override
    public void displayDefaultToolbarTitle() {
        setToolbarTitle(getString(R.string.photos));
    }

    @Override
    public void displayToolbarSubtitle(@Nullable PhotoAlbum album, @NonNull String text) {
        if (nonNull(album)) {
            setToolbarSubtitle(album.getDisplayTitle(requireActivity()) + " " + text);
        } else {
            setToolbarSubtitle(text);
        }
    }

    @Override
    public void scrollTo(int position) {
        mRecyclerView.scrollToPosition(position);
    }

    @Override
    public void setDrawerPhotosSelected(boolean selected) {
        if (requireActivity() instanceof OnSectionResumeCallback) {
            if (selected) {
                ((OnSectionResumeCallback) requireActivity()).onSectionResume(AbsNavigationFragment.SECTION_ITEM_PHOTOS);
            } else {
                ((OnSectionResumeCallback) requireActivity()).onClearSelection();
            }
        }
    }

    @Override
    public void returnSelectionToParent(List<Photo> selected) {
        Intent intent = new Intent();
        intent.putParcelableArrayListExtra(Extra.ATTACHMENTS, new ArrayList<>(selected));
        requireActivity().setResult(RESULT_OK, intent);
        requireActivity().finish();
    }

    @Override
    public void showSelectPhotosToast() {
        Toast.makeText(requireActivity(), getString(R.string.select_attachments), Toast.LENGTH_SHORT).show();
    }

    @Override
    public void startLocalPhotosSelection() {
        if (!AppPerms.hasReadStoragePermission(requireActivity())) {
            requestReadPermission.launch();
            return;
        }

        startLocalPhotosSelectionActibity();
    }

    @Override
    public void onCreateOptionsMenu(@NonNull Menu menu, @NonNull MenuInflater inflater) {
        super.onCreateOptionsMenu(menu, inflater);
        inflater.inflate(R.menu.menu_photos, menu);
    }

    @Override
    public void onPrepareOptionsMenu(@NonNull Menu menu) {
        super.onPrepareOptionsMenu(menu);
        menu.findItem(R.id.action_toggle_rev).setTitle(Settings.get().other().isInvertPhotoRev() ? R.string.sort_new_to_old : R.string.sort_old_to_new);
        menu.findItem(R.id.action_show_date).setVisible(!callPresenter(VkPhotosPresenter::getIsShowBDate, false));
    }

    @Override
    public void onToggleShowDate(boolean isShow) {
        mAdapter.setIsShowDate(isShow);
    }

    @Override
    public boolean onOptionsItemSelected(@NonNull MenuItem item) {
        if (item.getItemId() == R.id.action_get_downloaded) {
            if (!AppPerms.hasReadWriteStoragePermission(requireActivity())) {
                requestReadPermissionForLoadDownload.launch();
                return true;
            }
            callPresenter(VkPhotosPresenter::loadDownload);
            return true;
        } else if (item.getItemId() == R.id.action_show_date) {
            callPresenter(VkPhotosPresenter::doToggleDate);
            requireActivity().invalidateOptionsMenu();
        } else if (item.getItemId() == R.id.action_toggle_rev) {
            callPresenter(VkPhotosPresenter::togglePhotoInvert);
            requireActivity().invalidateOptionsMenu();
        }

        return super.onOptionsItemSelected(item);
    }

    private void startLocalPhotosSelectionActibity() {
        Intent intent = new Intent(requireActivity(), PhotosActivity.class);
        intent.putExtra(PhotosActivity.EXTRA_MAX_SELECTION_COUNT, Integer.MAX_VALUE);
        requestUploadPhoto.launch(intent);
    }

    @Override
    public void startLocalPhotosSelectionIfHasPermission() {
        if (AppPerms.hasReadStoragePermission(requireActivity())) {
            startLocalPhotosSelectionActibity();
        }
    }

    @NonNull
    @Override
    public IPresenterFactory<VkPhotosPresenter> getPresenterFactory(@Nullable Bundle saveInstanceState) {
        return () -> {
            ParcelableOwnerWrapper ownerWrapper = requireArguments().getParcelable(Extra.OWNER);
            Owner owner = nonNull(ownerWrapper) ? ownerWrapper.get() : null;
            PhotoAlbum album = requireArguments().getParcelable(Extra.ALBUM);

            return new VkPhotosPresenter(
                    requireArguments().getInt(Extra.ACCOUNT_ID),
                    requireArguments().getInt(Extra.OWNER_ID),
                    requireArguments().getInt(Extra.ALBUM_ID),
                    requireArguments().getString(Extra.ACTION, ACTION_SHOW_PHOTOS),
                    owner,
                    album,
                    saveInstanceState
            );
        };
    }
}

File: app_fenrir/src/main/kotlin/dev/ragnarok/fenrir/fragment/vkphotos/VKPhotosFragment.kt
package dev.ragnarok.fenrir.fragment.vkphotos

import android.Manifest
import android.app.Activity
import android.content.Intent
import android.os.Bundle
import android.view.*
import android.widget.TextView
import android.widget.Toast
import androidx.activity.result.ActivityResult
import androidx.activity.result.contract.ActivityResultContracts
import androidx.appcompat.app.AppCompatActivity
import androidx.core.view.MenuProvider
import androidx.recyclerview.widget.GridLayoutManager
import androidx.recyclerview.widget.RecyclerView
import androidx.swiperefreshlayout.widget.SwipeRefreshLayout
import com.google.android.material.floatingactionbutton.FloatingActionButton
import dev.ragnarok.fenrir.*
import dev.ragnarok.fenrir.activity.ActivityFeatures
import dev.ragnarok.fenrir.activity.PhotosActivity
import dev.ragnarok.fenrir.dialog.ImageSizeAlertDialog
import dev.ragnarok.fenrir.dialog.ImageSizeAlertDialog.Companion.showUploadPhotoSizeIfNeed
import dev.ragnarok.fenrir.fragment.base.BaseMvpFragment
import dev.ragnarok.fenrir.fragment.base.core.IPresenterFactory
import dev.ragnarok.fenrir.fragment.vkphotos.BigVkPhotosAdapter.UploadActionListener
import dev.ragnarok.fenrir.listener.EndlessRecyclerOnScrollListener
import dev.ragnarok.fenrir.listener.OnSectionResumeCallback
import dev.ragnarok.fenrir.listener.PicassoPauseOnScrollListener
import dev.ragnarok.fenrir.model.*
import dev.ragnarok.fenrir.model.wrappers.SelectablePhotoWrapper
import dev.ragnarok.fenrir.place.PlaceFactory.getPhotoAlbumGalleryPlace
import dev.ragnarok.fenrir.settings.Settings
import dev.ragnarok.fenrir.upload.Upload
import dev.ragnarok.fenrir.util.AppPerms.hasReadStoragePermission
import dev.ragnarok.fenrir.util.AppPerms.hasReadWriteStoragePermission
import dev.ragnarok.fenrir.util.AppPerms.requestPermissionsAbs
import dev.ragnarok.fenrir.util.ViewUtils.setupSwipeRefreshLayoutWithCurrentTheme
import dev.ragnarok.fenrir.util.toast.CustomToast
import dev.ragnarok.fenrir.view.navigation.AbsNavigationView

class VKPhotosFragment : BaseMvpFragment<VkPhotosPresenter, IVkPhotosView>(),
    BigVkPhotosAdapter.PhotosActionListener, UploadActionListener, IVkPhotosView, MenuProvider {
    private val requestUploadPhoto = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK) {
            val photos: ArrayList<LocalPhoto>? =
                result.data?.getParcelableArrayListExtraCompat(Extra.PHOTOS)
            if (photos.nonNullNoEmpty()) {
                onPhotosForUploadSelected(photos)
            }
        }
    }
    private val requestReadPermission =
        requestPermissionsAbs(
            arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
        ) { presenter?.fireReadStoragePermissionChanged() }
    private val requestReadPermissionForLoadDownload =
        requestPermissionsAbs(
            arrayOf(Manifest.permission.READ_EXTERNAL_STORAGE)
        ) { presenter?.loadDownload() }
    private val requestPhotoUpdate = registerForActivityResult(
        ActivityResultContracts.StartActivityForResult()
    ) { result: ActivityResult ->
        if (result.resultCode == Activity.RESULT_OK && result.data != null && (result.data
                ?: return@registerForActivityResult)
                .extras != null
        ) {
            lazyPresenter {
                updateInfo(
                    (result.data?.extras ?: return@lazyPresenter).getInt(Extra.POSITION),
                    ((result.data
                        ?: return@lazyPresenter)
                        .extras ?: return@lazyPresenter).getLong(Extra.PTR)
                )
            }
        }
    }
    private var mSwipeRefreshLayout: SwipeRefreshLayout? = null
    private var mAdapter: BigVkPhotosAdapter? = null
    private var mEmptyText: TextView? = null
    private var mFab: FloatingActionButton? = null
    private var mAction: String? = null
    private var mRecyclerView: RecyclerView? = null
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        mAction = requireArguments().getString(Extra.ACTION, IVkPhotosView.ACTION_SHOW_PHOTOS)
    }

    override fun onViewCreated(view: View, savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        requireActivity().addMenuProvider(this, viewLifecycleOwner)
    }

    override fun onCreateView(
        inflater: LayoutInflater,
        container: ViewGroup?,
        savedInstanceState: Bundle?
    ): View? {
        val root = inflater.inflate(R.layout.fragment_photo_gallery, container, false)
        (requireActivity() as AppCompatActivity).setSupportActionBar(root.findViewById(R.id.toolbar))
        val columnCount = resources.getInteger(R.integer.local_gallery_column_count)
        val manager: RecyclerView.LayoutManager = GridLayoutManager(requireActivity(), columnCount)
        mSwipeRefreshLayout = root.findViewById(R.id.refresh)
        mSwipeRefreshLayout?.setOnRefreshListener {
            presenter?.fireRefresh()
        }
        setupSwipeRefreshLayoutWithCurrentTheme(requireActivity(), mSwipeRefreshLayout)
        mRecyclerView = root.findViewById(R.id.list)
        mRecyclerView?.layoutManager = manager
        mRecyclerView?.addOnScrollListener(PicassoPauseOnScrollListener(TAG))
        mRecyclerView?.addOnScrollListener(object : EndlessRecyclerOnScrollListener() {
            override fun onScrollToLastElement() {
                presenter?.fireScrollToEnd()
            }
        })
        mEmptyText = root.findViewById(R.id.empty)
        mFab = root.findViewById(R.id.fr_photo_gallery_attach)
        mFab?.setOnClickListener { onFabClicked() }
        mAdapter = BigVkPhotosAdapter(requireActivity(), emptyList(), emptyList(), TAG)
        mAdapter?.setPhotosActionListener(this)
        mAdapter?.setUploadActionListener(this)
        mRecyclerView?.adapter = mAdapter
        return root
    }

    private fun resolveEmptyTextVisibility() {
        if (mEmptyText != null && mAdapter != null) {
            mEmptyText?.visibility =
                if (mAdapter?.itemCount == 0) View.VISIBLE else View.GONE
        }
    }

    private fun resolveFabVisibility(show: Boolean) {
        if (!isAdded || mFab == null) return
        if (mFab?.isShown == true && !show) {
            mFab?.hide()
        }
        if (mFab?.isShown == false && show) {
            mFab?.show()
        }
    }

    private fun onFabClicked() {
        if (isSelectionMode) {
            presenter?.fireSelectionCommitClick()
        } else {
            presenter?.fireAddPhotosClick()
        }
    }

    private val isSelectionMode: Boolean
        get() = IVkPhotosView.ACTION_SELECT_PHOTOS == mAction

    private fun onPhotosForUploadSelected(photos: List<LocalPhoto>) {
        showUploadPhotoSizeIfNeed(
            requireActivity(),
            object : ImageSizeAlertDialog.Callback {
                override fun onSizeSelected(size: Int) {
                    doUploadPhotosToAlbum(photos, size)
                }
            })
    }

    internal fun doUploadPhotosToAlbum(photos: List<LocalPhoto>, size: Int) {
        presenter?.firePhotosForUploadSelected(
            photos,
            size
        )
    }

    override fun onResume() {
        super.onResume()
        ActivityFeatures.Builder()
            .begin()
            .setHideNavigationMenu(false)
            .setBarsColored(requireActivity(), true)
            .build()
            .apply(requireActivity())
    }

    override fun onDestroyView() {
        mAdapter?.cleanup()
        super.onDestroyView()
    }

    override fun onPhotoClick(
        holder: BigVkPhotosAdapter.PhotoViewHolder,
        photoWrapper: SelectablePhotoWrapper
    ) {
        if (isSelectionMode) {
            presenter?.firePhotoSelectionChanged(
                photoWrapper
            )
            mAdapter?.updatePhotoHoldersSelectionAndIndexes()
        } else {
            presenter?.firePhotoClick(
                photoWrapper
            )
        }
    }

    override fun onUploadRemoveClicked(upload: Upload) {
        presenter?.fireUploadRemoveClick(
            upload
        )
    }

    override fun displayData(photos: List<SelectablePhotoWrapper>, uploads: List<Upload>) {
        if (mAdapter != null) {
            mAdapter?.setData(BigVkPhotosAdapter.DATA_TYPE_UPLOAD, uploads)
            mAdapter?.setData(BigVkPhotosAdapter.DATA_TYPE_PHOTO, photos)
            mAdapter?.notifyDataSetChanged()
            resolveEmptyTextVisibility()
        }
    }

    override fun notifyDataSetChanged() {
        if (mAdapter != null) {
            mAdapter?.notifyDataSetChanged()
            resolveEmptyTextVisibility()
        }
    }

    override fun notifyPhotosAdded(position: Int, count: Int) {
        if (mAdapter != null) {
            mAdapter?.notifyItemRangeInserted(position, count, BigVkPhotosAdapter.DATA_TYPE_PHOTO)
            resolveEmptyTextVisibility()
        }
    }

    override fun notifyPhotosChanged(position: Int, count: Int) {
        if (mAdapter != null) {
            mAdapter?.notifyItemRangeChanged(position, count)
            resolveEmptyTextVisibility()
        }
    }

    override fun displayRefreshing(refreshing: Boolean) {
        mSwipeRefreshLayout?.isRefreshing = refreshing
    }

    override fun notifyUploadAdded(position: Int, count: Int) {
        if (mAdapter != null) {
            mAdapter?.notifyItemRangeInserted(position, count, BigVkPhotosAdapter.DATA_TYPE_UPLOAD)
            resolveEmptyTextVisibility()
        }
    }

    override fun notifyUploadRemoved(index: Int) {
        if (mAdapter != null) {
            mAdapter?.notifyItemRemoved(index, BigVkPhotosAdapter.DATA_TYPE_UPLOAD)
            resolveEmptyTextVisibility()
        }
    }

    override fun setButtonAddVisible(visible: Boolean, anim: Boolean) {
        if (mFab != null) {
            resolveFabVisibility(visible)
        }
    }

    override fun notifyUploadItemChanged(index: Int) {
        mAdapter?.notifyItemChanged(index, BigVkPhotosAdapter.DATA_TYPE_UPLOAD)
    }

    override fun notifyUploadProgressChanged(id: Int, progress: Int) {
        mAdapter?.updateUploadHoldersProgress(id, true, progress)
    }

    override fun displayGallery(
        accountId: Long,
        albumId: Int,
        ownerId: Long,
        source: TmpSource,
        position: Int
    ) {
        getPhotoAlbumGalleryPlace(
            accountId,
            albumId,
            ownerId,
            source,
            position,
            false,
            Settings.get().other().isInvertPhotoRev
        ).tryOpenWith(requireActivity())
    }

    override fun displayGalleryUnSafe(
        accountId: Long,
        albumId: Int,
        ownerId: Long,
        parcelNativePointer: Long,
        position: Int
    ) {
        getPhotoAlbumGalleryPlace(
            accountId,
            albumId,
            ownerId,
            parcelNativePointer,
            position,
            false,
            Settings.get().other().isInvertPhotoRev
        ).setActivityResultLauncher(requestPhotoUpdate).tryOpenWith(requireActivity())
    }

    override fun displayDefaultToolbarTitle() {
        setToolbarTitle(getString(R.string.photos))
    }

    override fun displayToolbarSubtitle(album: PhotoAlbum?, text: String) {
        if (album != null) {
            setToolbarSubtitle(album.getDisplayTitle(requireActivity()) + " " + text)
        } else {
            setToolbarSubtitle(text)
        }
    }

    override fun scrollTo(position: Int) {
        mRecyclerView?.scrollToPosition(position)
    }

    override fun setDrawerPhotosSelected(selected: Boolean) {
        if (requireActivity() is OnSectionResumeCallback) {
            if (selected) {
                (requireActivity() as OnSectionResumeCallback).onSectionResume(AbsNavigationView.SECTION_ITEM_PHOTOS)
            } else {
                (requireActivity() as OnSectionResumeCallback).onClearSelection()
            }
        }
    }

    override fun returnSelectionToParent(selected: List<Photo>) {
        val intent = Intent()
        intent.putParcelableArrayListExtra(Extra.ATTACHMENTS, ArrayList(selected))
        requireActivity().setResult(Activity.RESULT_OK, intent)
        requireActivity().finish()
    }

    override fun showSelectPhotosToast() {
        CustomToast.createCustomToast(requireActivity()).setDuration(Toast.LENGTH_LONG)
            .showToastInfo(R.string.select_attachments)
    }

    override fun startLocalPhotosSelection() {
        if (!hasReadStoragePermission(requireActivity())) {
            requestReadPermission.launch()
            return
        }
        startLocalPhotosSelectionActivity()
    }

    override fun onCreateMenu(menu: Menu, menuInflater: MenuInflater) {
        menuInflater.inflate(R.menu.menu_photos, menu)
    }

    override fun onPrepareMenu(menu: Menu) {
        super.onPrepareMenu(menu)
        menu.findItem(R.id.action_toggle_rev).setTitle(
            if (Settings.get()
                    .other().isInvertPhotoRev
            ) R.string.sort_new_to_old else R.string.sort_old_to_new
        )
        menu.findItem(R.id.action_show_date).isVisible =
            !(presenter?.isShowBDate ?: false)
    }

    override fun onToggleShowDate(isShow: Boolean) {
        mAdapter?.setIsShowDate(isShow)
    }

    override fun onMenuItemSelected(menuItem: MenuItem): Boolean {
        when (menuItem.itemId) {
            R.id.action_get_downloaded -> {
                if (!hasReadWriteStoragePermission(requireActivity())) {
                    requestReadPermissionForLoadDownload.launch()
                    return true
                }
                presenter?.loadDownload()
                return true
            }
            R.id.action_show_date -> {
                presenter?.doToggleDate()
                requireActivity().invalidateOptionsMenu()
                return true
            }
            R.id.action_toggle_rev -> {
                presenter?.togglePhotoInvert()
                requireActivity().invalidateOptionsMenu()
                return true
            }
        }
        return false
    }

    private fun startLocalPhotosSelectionActivity() {
        val intent = Intent(requireActivity(), PhotosActivity::class.java)
        intent.putExtra(PhotosActivity.EXTRA_MAX_SELECTION_COUNT, Int.MAX_VALUE)
        requestUploadPhoto.launch(intent)
    }

    override fun startLocalPhotosSelectionIfHasPermission() {
        if (hasReadStoragePermission(requireActivity())) {
            startLocalPhotosSelectionActivity()
        }
    }

    override fun getPresenterFactory(saveInstanceState: Bundle?): IPresenterFactory<VkPhotosPresenter> {
        return object : IPresenterFactory<VkPhotosPresenter> {
            override fun create(): VkPhotosPresenter {
                val ownerWrapper: ParcelableOwnerWrapper? =
                    requireArguments().getParcelableCompat(Extra.OWNER)
                val owner = ownerWrapper?.get()
                val album: PhotoAlbum? = requireArguments().getParcelableCompat(Extra.ALBUM)
                return VkPhotosPresenter(
                    requireArguments().getLong(Extra.ACCOUNT_ID),
                    requireArguments().getLong(Extra.OWNER_ID),
                    requireArguments().getInt(Extra.ALBUM_ID),
                    requireArguments().getString(Extra.ACTION, IVkPhotosView.ACTION_SHOW_PHOTOS),
                    owner,
                    album,
                    requireArguments().getInt(Extra.SELECTED),
                    saveInstanceState
                )
            }
        }
    }

    companion object {
        private val TAG = VKPhotosFragment::class.java.simpleName
        fun buildArgs(
            accountId: Long,
            ownerId: Long,
            albumId: Int,
            action: String?,
            selected: Int
        ): Bundle {
            val args = Bundle()
            args.putLong(Extra.ACCOUNT_ID, accountId)
            args.putLong(Extra.OWNER_ID, ownerId)
            args.putInt(Extra.ALBUM_ID, albumId)
            args.putInt(Extra.SELECTED, selected)
            args.putString(Extra.ACTION, action)
            return args
        }

        fun newInstance(args: Bundle?): VKPhotosFragment {
            val fragment = VKPhotosFragment()
            fragment.arguments = args
            return fragment
        }

        fun newInstance(
            accountId: Long,
            ownerId: Long,
            albumId: Int,
            action: String?
        ): VKPhotosFragment {
            return newInstance(buildArgs(accountId, ownerId, albumId, action, -1))
        }
    }
}